### DISCUSSION TITLE
Using assetPrefix in statically generated pages

### DISCUSSION THREAD

**Discussion by ussuritiger** [vercel/next.js#14119] [Help]
My app is using custom server and has both server-rendered and statically generated pages. The site has two environments: acceptance and production. While building the application (once) I also do static HTML export (twice per environment) in order to upload static assets into CDN.

The CDN URL is used then in the runtime as `assetPrefix` and can be configured in `next.config.js` per environment. It works well for server-rendered pages but the statically generated pages do not use that `assetPrefix` and as result all the static assets are referring to root (`/`) and not accessible in my case.

To be more clear, the `assetPrefix` is specified only for `PHASE_EXPORT` and `PHASE_PRODUCTION_SERVER`.

Therefore I am wondering whether it is possible to use `assetPrefix` in runtime (`PHASE_PRODUCTION_SERVER` phase) for statically generated pages without building the app two times (the app is dockerized so building two separate images feels quite sub-optimal).

**Comment by xian0831** [comment]
I am running into the similar issue. Asset prefix, CDN path in my case, is added to all JS/CSS assets for server rendered pages. But it doesn't work on the statically generated pages. 

I try both true & false on the fallback options but it returns the same result.

**Comment by pcardune** [comment]
I don't think it's possible to use an `assetPrefix` that changes at runtime in conjunction with statically generated pages. This should really be documented in the [CDN support docs](https://nextjs.org/docs/api-reference/next.config.js/cdn-support-with-asset-prefix). There is a note about something similar on the [Runtime Configuration](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration) page:

> Generally you'll want to use _*build-time*_ environment variables to provide your configuration. The reason for this is that runtime configuration adds rendering / initialization overhead and is _*incompatible with Automatic Static Optimization*_.

I think the only way to deal with this is to run `next build` inside your docker image when it starts up.

**Comment by mlstubblefield** [comment]
This is a problem for us as well when it comes to having multiple static nextjs apps on the same domain.
![image](https://user-images.githubusercontent.com/7562533/117517542-eb577d00-af61-11eb-903d-11ee368551a8.png)

We tried using base path as well as it didn't seem to have this specific issue, but when we went down that route when calling router methods, the URLs were still including the basePath when we wanted to exclude it (we get rid of the base path on the server for the actual URL). I tried to use the "as" parameter, but it looks like as also includes base path.
Because we only want the base path on the static resources (js, css, _next directories), I felt like assetPrefix would be the better option honestly, but the screenshot at the top made it not possible.
Below are some screenshots describing the issues we had with base path.
![image](https://user-images.githubusercontent.com/7562533/117518020-d380f880-af63-11eb-9d7d-85019b73fd71.png)
![image](https://user-images.githubusercontent.com/7562533/117518194-720d5980-af64-11eb-9bea-c36a48f85f02.png)
![image](https://user-images.githubusercontent.com/7562533/117518210-7df91b80-af64-11eb-9ab8-1606323ec006.png)

**Reply by mlstubblefield** [reply]
So for our specific use case **either** the following should fix our problem. I have no preference.
- Add an option to react router push/replace to ignore bsae path (perhaps utilizing the as paramter)
- Utilize assetPrefix for everything under the _next path including /_next/data/ requests.

**Comment by timur-hilmutdinov** [comment]
https://github.com/vercel/next.js/discussions/25681#discussioncomment-2026813 - fresher issue with a workaround.
